torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -172,6 +172,76 @@ std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
   return extent_to_min_input_id_extent;
 }
 
+bool shouldFuseBroadCast(const BroadcastOp* broadcast) {
+  auto broadcast_inp = broadcast->in()->as<TensorView>();
+
+  // skip if the broadcast directly uses fusion input,
+  //  meaning there's no reduction to be fused with.
+  if (broadcast_inp->isFusionInput()) {
+    return false;
+  }
+
+  // Check that the producer of the input is a warp reduction,
+  //  which could fuse with block broadcast on threadIdx.x
+  if (auto reduction_op =
+          dynamic_cast<ReductionOp*>(broadcast_inp->definition())) {
+    if (!ir_utils::getMaybeWarpReductionDim(reduction_op).has_value()) {
+      return false;
+    }
+  } else {
+    return false;
+  }
+
+  auto broadcast_out = broadcast->out()->as<TensorView>();
+
+  // Only support fusing register to register operations
+  //  at this stage.
+  if (broadcast_inp->getMemoryType() != MemoryType::Local ||
+      broadcast_out->getMemoryType() != MemoryType::Local) {
+    return false;
+  }
+
+  // Check that the broadcast operation is only either serial
+  //  or across threadIdx.x of the block.
+  int broadcast_pos = -1;
+  for (size_t i = 0; i < broadcast_out->nDims(); i++) {
+    if (broadcast_out->axis(i)->isBroadcast()) {
```

Comment:
`isBroadcast()` and `isParallelized()` don't necessarily mean an axis is translated to a parallel broadcast op. See `ThreadPredicateMap::getParallelBroadcastDomains()`. If a thread dimension is not predicated, it should be fine to ignore that axis.

Reply:
Thanks for pointing this out! I have updated this part to check the pred maps.

Reply:
This is already checking TIDx parallelization on the reduction operation followed by the broadcast. This guarantees a parallel broadcast. Maybe I'm just 22 days late to this so maybe it changed but we don't need to check the thread predicate map. @naoyam for confirmation.

Reply:
@csarofeen is right. It's guaranteed here.